### ISSUE TITLE
Sync animation in BlendSpace1D/2D

### ISSUE THREAD

**Issue by Kinwailo** [godotengine/godot#23414]
**Godot version:**
3.1 alpha1 (https://downloads.tuxfamily.org/godotengine/3.1/alpha1/)

**OS/device including version:**
Win7

**Issue description:**
Blending animation in BlendSpace1D/2D seem not synced and weird. I don't know how Godot blend the animation, I think if using two animation with similar action (footstep) with different speed, the blend position taken from two source animation should be the percent of time, so the different of length in two source animation does not affect the sync of final animation.

Sorry for my english.

**Steps to reproduce:**
Prepare two animation (run and run faster)
Create AnimationTree and add BlendSpace1D
Add run in 0 and run faster in 1
Click the cursor in 0.5

![2018-10-31 11-17-59](https://user-images.githubusercontent.com/28286958/47764381-d306a780-dcff-11e8-94da-c3faf22f9fe4.gif)

**Comment by Kinwailo** [comment]
I try to do a simulation in blender, I prepared a walk and a run animation:
![walk](https://user-images.githubusercontent.com/28286958/47893785-d03fba00-de99-11e8-81c6-acdd987ea9a2.gif)
![run](https://user-images.githubusercontent.com/28286958/47893790-d33aaa80-de99-11e8-9551-3bd132ab5970.gif)

I add two animation in nla editor and blend in 50%:
![1](https://user-images.githubusercontent.com/28286958/47893814-f2d1d300-de99-11e8-9e1a-f9a3f0aecf90.jpg)
The result is similar to godot.
![weird](https://user-images.githubusercontent.com/28286958/47893826-067d3980-de9a-11e8-9b1f-8e82818b68ff.gif)

If I scale two animation in same length, the result is prefect.
![2](https://user-images.githubusercontent.com/28286958/47893890-6a076700-de9a-11e8-8b19-985a7d814eb5.jpg)
![correct](https://user-images.githubusercontent.com/28286958/47893857-39bfc880-de9a-11e8-9969-3957ce6274e1.gif)

So, if the blendspace do correctly, I can have different walk speed with different blend factor.

**Comment by Jackain** [comment]
I'm struggling as well to synch two animation (walk-cycle and run-cycle)

**Comment by David-Ochoa** [comment]
I've seen in other engines they scale the longest animation to match the shortest, but sometimes it's not enough and other adjustments need to be applied. In my opinion a good solution would be to add a scale property to the AnimationNodeAnimation node.
Right now I'm scaling the animation before importing and it works fine, but having the option to change it in the editor would be better.

**Comment by David-Ochoa** [comment]
Well, after writing the last comment I started to find a way to scale the animation, and the solution is using a  BlendTree inside the BlendSpace1D/2D node, I recorded a video where the process is explained:
[Fix blending](https://youtu.be/13S76dOIaDc)

**Comment by clayjohn** [comment]
@Kinwailo did David-Ochoa's solution fix your problem?

**Comment by AdamRatai** [comment]
I just bumped on this problem. It is something really wrong, I think, because I have few animations not only with exact amount of frames (100 per each animation) but I even synced them so legs are more less in the same position in same frame (I mean left leg is for example in front in 15 frame in every animation). But when I use Blendspace 1D - they make exactly the same weird inbetweens as I see at provided gifs.
I will check David-Ochoa solution, but even if it will work for me, I found it a workaround and I think it should be fixed, or explained by author.

**Comment by David-Ochoa** [comment]
I have the issue again, but this time is for three animations. I have Idle, Walk and Run. It starts correctly but if the point goes from Idle to walk/run sometimes the animation loops awkwardly. It happens using Godot's scale node or scaling before importing: [Test](https://youtu.be/WOi5Kb35IIA)

**Comment by n-pigeon** [comment]
Yeah, I had this problem too. Animations have to be same length for the blend to work which sucks a lot and is unintuitive why it is broken in the first place :(

**Comment by Shadowblitz16** [comment]
I think I just had this problem too.
I was blending a walking and running animation where the running animation was shorter and faster

**Comment by d3is** [comment]
Been dealing with this bug for a while now and thought it was just something wrong with my animations. So happy to find this as a confirmed bug now.

**Comment by TokageItLab** [comment]
This problem is forcing me to use `TransitionNode` and `SeekNode` without using `BlendSpace2DNode`. Also, there is a lack of implementation for most node; we will have to improve the feature of the animation nodes a lot before we get to Godot 4.0.

**Comment by HungryProton** [comment]
So, just in case #34134 and  #34179 never get merged, did anyone found a workaround for this? Using a timescale node as suggested by David-Ochoa still desyncs the animations after a while.

**Comment by TokageItLab** [comment]
@HungryProton This is a complete hack, but you can always force synchronization by directly setting the value to NodeSeek. However, if you use root motion, you will need a PR #41728. While, since NodeSeek doesn't support looping root motion, so you need to let the engine do the looping the animation when the frame is over the animation length with stopping your seeking.

**Comment by HungryProton** [comment]
@TokageItLab Thanks for the reply!
That was more convoluted than I thought, but I finally got something working. Using two NodeSeek and a bit of GDScript seems to do the trick for now.

![image](https://user-images.githubusercontent.com/52043844/104141648-1e28a480-53b8-11eb-9423-4f857f024cc1.png)
(In this screenshot, the Blend2 blend amount (`move_blend` variable) is calculated based on my character's speed).

**Comment by timkrief** [comment]
A "fix" for this is having all the animations at the same length and then putting them all in BlendTrees to add a TimeScale node. That way you can blend them together perfectly in a BlendSpace and change the overall speed by tweaking all the time scale parameters all at once.

**Comment by AdamRatai** [comment]
Maybe something changed, but thats exactly what was NOT WORKING in my case.

**Comment by timkrief** [comment]
> Maybe something changed, but thats exactly what was NOT WORKING in my case.

You're right, it seemed to work at first but for some reason it stills desync ...

**Comment by TokageItLab** [comment]
Even if you use a Node that is not a BlendSpace, you must always set Sync to True in NodeBlend and NodeAdd for animations that you want to synchronize, but there are some Nodes that do not have Sync, such as NodeTransition. The current solution is to implement your own timer in GDScript and run all animations with that timer and NodeSeek.

**Comment by AdamRatai** [comment]
Thanks! In next attempt I will try it. But i hope it will be fixed some day ;).

**Comment by krm01** [comment]
I was banging my head against this for a while before finding the thread here, Godot version 3.2.3. Tried all of the suggestions and nothing worked.

Fed up, I decided to just accept it for now, and started up my game. To my horror... it looked fine. At least in my case, this seems to be a bug only with the editor viewport, and isn't manifesting in-game. Maybe my animations are simpler, I don't know. But I can live with the viewport bug. I won't get these hours back but maybe someone else in my situation will 🙃

**Comment by TokageItLab** [comment]
For example, sideways walking, a simple blend of strafing and back-and-forth walking like in the [tps-demo](https://godotengine.org/asset-library/asset/678), will not cause problems, but cases where you have to turn your hips will cause problems because of the synchronization required.

**Comment by TokageItLab** [comment]
@and-rad `p_optimize` may be a bad name. It is actually there to pass p_sync to the chained parent node.

To consider about optimization: in the Skeleton case, it is obvious that we need to blend stopped and playing animations, and in cases like the Audio and Method tracks, we need to blend them because bugs can occur at the end of the timeline. Method track has a bug and I sent a PR #61885 to fix it last time.

Your understanding is correct, the `p_optimize` parameter only works to cause asynchrony if it is not necessary to blend completely.

I agree that this feature is necessary. Without it, it would be difficult to create a TPS game. As I mentioned the solution above, the combination of Sync and TimeScale can actually solve this problem as an alternative to #34179. Moreover, I also think that a Sync parameter should be added to NodeTransition.

**Comment by TokageItLab** [comment]
Do you mean implementing Sync and TimeScale on all `AnimationNode`s instead of `AnimationNodeAnimation`? Sounds good to me, but I am a little concerned about the effects of this. Well, I think that TimeScale could be removed in that case.
It might be better to discuss this in the contributor chat with animation team.

**Comment by and-rad** [comment]
> Do you mean implementing Sync and TimeScale on all `AnimationNode`s instead of `AnimationNodeAnimation`?

Yes, exactly. At least for Sync. I'm not sure yet about TimeScale.

**Comment by TokageItLab** [comment]
I believe that every node should have Sync.However, I feel that for TimeScale, it may be okay to leave it as an individual node.
It would be like having NodeTimeScale chained to every node child. However, we need to look into what the fact that NodeTimeScale sets sync to false is affecting.

**Comment by RetroBum** [comment]
Just had this issue in 3.5 and quickly tried the exact same model/animations in 4.0 and had the exact same issue with or without sync option enabled. It didn't seem to make a difference at all. The only solution, in 3.5 and 4.0, was editing the animations to the be the exact same length inside the AnimationPlayer.

**Comment by TokageItLab** [comment]
See https://github.com/godotengine/godot/pull/62623. The workaround of synchronizing animations of different lengths by normalizing the proportions by TimeScale works well.

To implement that normalization automation into the core, the current blending system is not guaranteed to be sufficient to cover all cases.

However, it should be possible to script the automatic configuration of common case-specific animation combinations such as TPS and FPS as an add-on.

BTW in 3.5, BlendSpace does not have sync, so that workaround cannot be used.

**Comment by Maveyyl** [comment]
On 4.1.beta1.mono I don't feel like sync is doing what is should be doing. I have 5 animations of same length: neutral and all 4 direction walk animations. If I repeatedly move the blend position from 1:0 to 1:1 I get different animations every time for the 1:1 position, doesn't seem that the "walk forward" and "walk right" animations actually synchronized.

**Comment by TokageItLab** [comment]
@Maveyyl I remember that sync was temporarily broken in 4.3.dev, but I don't remember any problems in 4.1. In any case, there was a major rework on time progression in 4.3, so I recommend trying 4.3 or later.

**Comment by Maveyyl** [comment]
> @Maveyyl I remember that sync was temporarily broken in 4.3.dev, but I don't remember any problems in 4.1. In any case, there was a major rework on time progression in 4.3, so I recommend trying 4.3 or later.

Oops I meant 4.3.beta1 it's a typo sorry. So you confirm it's normal and I should wait?

**Comment by TokageItLab** [comment]
@Maveyyl It has been fixed by https://github.com/godotengine/godot/pull/92739, so that fix will be included in beta2.

**Comment by Biran0** [comment]
In Godot 4.3. I blend very nice Walking animation with strafing animations, while the enemy is walking ..but when I tried to do the same thing while crouching it looks very bad. The mixamo animations I tried are perfectly synced, I checked them frame by frame.

However, the enemy sometimes looks great but sometimes its like the sync doesnt work at all, it looks like enemy tries to lift both legs at the same time from the ground. Makes something like he is trying to hop.
(I mean I have the enemy look at the player character and walk to different directions, and so he goes in sync or out of sync randomly while he is changing directions)

So, what I want to say is that the sync of the mixamo animations for sure is fine.. but Im not sure about the sync of Godot. I have it checked but not sure if it works.
